### PR TITLE
Compiler: rewrite the js parser using menhir incremental api

### PR DESCRIPTION
**Pull request by hhugo**
Before this PR, the js lexer was trying to adjust its production based on the previous token.

- INCR vs INCR_NB ("NB" stands for "No line Break")
- DECR vs DECR_NB
- DIV vs DIV_ASSIGN vs REGEXP

The lexer was also populating the fol (first of line) field that was later used by `Js_parser.Lexer.adjust_token` and the js parser for semicolon automatic insertion.

Using the incremental API of menhir (where the control flow is inverted), we can now remove all theses hacks and move all the logic inside `parser_aux`.

This also has the advantage of being much more friendly with the stack when running in javascript (jsoo toplevel)

### CODE REVIEW

**Comment by hhugo** [comment]
fyi @let-def

**Comment by let-def** [comment]
Cool!

**Comment by TyOverby** [comment]
Is the goal with this PR to allow for better regexp parsing?

File: compiler/lib/js_lexer.mll [hunk 1]
```diff
@@ -74,16 +74,7 @@ let update_loc lexbuf ?file ~line ~absolute chars =
     pos_bol = pos.pos_cnum - chars;
                               }
 
-let tokinfo prev lexbuf =
```
[the remark this answers is not in the document]
Comment (hhugo) [reply compiler/lib/js_lexer.mll:77]: `tokinfo` was using `prev` to populate the `fol` field which was later used elsewhere for deal with semi-colon automatic insertion. I've updated the PR description.

**Comment by hhugo** [comment]
This PR allows us to parse regexp literal correctly and in a much cleaner way.

File: compiler/lib/parse_js.ml [hunk 2]
```diff
@@ -19,115 +19,165 @@
...
+                    T_VIRTUAL_SEMICOLON Parse_info.zero
+                (* The practical effect of these restricted productions is as follows:
+                 * When a ++ or -- token is encountered where the parser would treat it
+                 * as a postfix operator, and at least one LineTerminator occurred between
+                 * the preceding token and the ++ or -- token, then a semicolon is automatically
+                 * inserted before the ++ or -- token. *)
+                | _, (T_DECR cpi as tok) when not (fol prev tok) -> Js_token.T_DECR_NB cpi
+                | _, (T_INCR cpi as tok) when not (fol prev tok) -> Js_token.T_INCR_NB cpi
+                | _, ((T_DIV _ | T_DIV_ASSIGN _) as tok) ->
+                    if I.acceptable checkpoint tok lexbuf.Lexing.lex_start_p
```
Comment (TyOverby) [inline compiler/lib/parse_js.ml:107]: What does this condition mean?

Comment (hhugo) [reply compiler/lib/parse_js.ml:107]: This is part of the incremental api. It checks if a token is valid given a parser state (checkpoint).
Div and Regexp tokens and mutually exclusive in the JavaScript parser. We first check if div is an acceptable token, if not, we reset the lexer and lex a Regexp instead

**Pull request merged**